x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/PushTopNToSource.java

```diff
@@ -7,55 +7,204 @@
-            return queryExec.canPushSorts();
+    }
+
+    private static Pushable evaluatePushable(TopNExec topNExec, Predicate<FieldAttribute> hasIdenticalDelegate) {
+        PhysicalPlan child = topNExec.child();
+        if (child instanceof EsQueryExec queryExec
+            && queryExec.canPushSorts()
+            && canPushDownOrders(topNExec.order(), hasIdenticalDelegate)) {
+            // With the simplest case of `FROM index | SORT ...` we only allow pushing down if the sort is on a field
+            return new PushableQueryExec(queryExec);
+        }
+        if (child instanceof ExchangeExec exchangeExec
+            && exchangeExec.child() instanceof EsQueryExec queryExec
+            && queryExec.canPushSorts()
+            && canPushDownOrders(topNExec.order(), hasIdenticalDelegate)) {
+            // When we have an exchange between the FROM and the SORT, we also only allow pushing down if the sort is on a field
+            return new PushableExchangeExec(exchangeExec, queryExec);
+        }
+        if (child instanceof EvalExec evalExec && evalExec.child() instanceof EsQueryExec queryExec && queryExec.canPushSorts()) {
+            // When we have an EVAL between the FROM and the SORT, we consider pushing down if the sort is on a field and/or
+            // a distance function defined in the EVAL. We also move the EVAL to after the SORT.
+            List<Order> orders = topNExec.order();
+            List<Alias> fields = evalExec.fields();
+            LinkedHashMap<NameId, StDistance> refs = new LinkedHashMap<>();
+            LinkedHashMap<NameId, FieldAttribute> refAttributes = new LinkedHashMap<>();
+            fields.forEach(alias -> {
+                // TODO: can we support CARTESIAN also?
+                if (alias.child() instanceof StDistance distance && distance.crsType() == BinarySpatialFunction.SpatialCrsType.GEO) {
+                    refs.put(alias.id(), distance);
+                } else if (alias.child() instanceof FieldAttribute fieldAttribute) {
+                    refAttributes.put(alias.id(), fieldAttribute);
+                } else if (alias.child() instanceof ReferenceAttribute ref && refAttributes.containsKey(ref.id())) {
+                    FieldAttribute fieldAttribute = refAttributes.get(ref.id());
+                    refAttributes.put(alias.id(), fieldAttribute);
+                }
+            });
+            List<EsQueryExec.Sort> pushableSorts = new ArrayList<>();
+            for (Order order : orders) {
+                if (LucenePushDownUtils.isPushableFieldAttribute(order.child(), hasIdenticalDelegate)) {
+                    pushableSorts.add(
+                        new EsQueryExec.FieldSort(
+                            ((FieldAttribute) order.child()).exactAttribute(),
+                            order.direction(),
+                            order.nullsPosition()
+                        )
+                    );
```

Comment:
See my previous comment, pick on the sort then move backwrds - instead of expecting an eval then a sort.
that might be common scenario however there might be minor variations that could mke this to fail.

Reply:
I'm not sure I follow. This class seems to do what you suggest. Start with the sort (TopNExec) and then look backwards (to child->child, etc.). I do differentiate between a simple TopNExec->EsQueryExec, which is sufficient for a sort-by-field query, and a more advanced TopNExec->EvalExec->EsQueryExec, which is *required* for the distance sort because we cannot sort on expressions, only attributes. This PR does, however, allow sorting on trivial aliases too, like `FROM index | EVAL x = y | SORT x | LIMIT 10`.

Reply:
As far as I can see, this PR only expands the range of queries that can be pushed down. Each time you describe a case that might `fail`, it seems more like a case that might not get the optimization. I do not (yet) see an actual failure.